chore: Move token verification logic to util

lahirumaramba opened the pull request:
- Move `FirebaseTokenVerifier` to util
- Include error type and codes to `FirebaseTokenInfo` to throw public error types from `FirebaseTokenVerifier`

hiranya911 commented on src/utils/token-verifier.ts:
```diff
@@ -43,6 +44,8 @@ export const ID_TOKEN_INFO: FirebaseTokenInfo = {
   jwtName: 'Firebase ID token',
   shortName: 'ID token',
   expiredErrorCode: AuthClientErrorCode.ID_TOKEN_EXPIRED,
+  errorCodeType: AuthClientErrorCode,
```
ID_TOKEN_INFO and SESSION_COOKIE_INFO should be defined in the auth module.

lahirumaramba replied:
Good call. Moved all auth related items to a new `token-verifier-util.ts`

hiranya911 commented on src/utils/error.ts:
```diff
@@ -338,10 +338,28 @@ export class AppErrorCodes {
+/**
+ * Base class for client error codes and their default messages.
+ */
+export class BaseClientErrorCode {
```
This is a fairly weak abstraction. It's a class with no specific shape, but a few static members. I think this ought to be an interface:

```typescript
interface ErrorCodeConfig {
  invalidArg: string;
  invalidCredential: string;
  internalError: string;
}
```

WDYT?

lahirumaramba replied:
Agreed! Given the number of error codes needed in the token verifier util this makes sense. Thanks!

hiranya911 reviewed (changes requested):
Haven't looked at everything, but looks reasonable. Let's start off with some feedback on the high-level design.

lahirumaramba commented on src/auth/token-verifier-util.ts:
```diff
@@ -0,0 +1,89 @@
+  jwtName: 'Firebase ID token',
+  shortName: 'ID token',
+  expiredErrorCode: AuthClientErrorCode.ID_TOKEN_EXPIRED,
+  errorCodeConfig: AUTH_ERROR_CODE_CONFIG,
```
I am debating if we should include `expiredErrorCode` in `errorCodeConfig` or not. If we do so `ErrorCodeConfig` interface will look more of a token related config. WDYT?

hiranya911 replied:
I expected it to be a token verification specific thing. I don't think it has any use outside of that context. +1 for merging the expiredErrorCode into the same interface.

hiranya911 commented on src/auth/token-verifier-util.ts:
```diff
@@ -0,0 +1,89 @@
+/*!
```
These can go in the token-verifier module like they used to. No need to introduce a new module.

lahirumaramba replied:
Some of the components `ID_TOKEN_INFO` and `SESSION_COOKIE_INFO` are `auth` specific. Instead of moving them to the `auth` module, I introduced the token verifier util module for better separation. Would you rather have them all in the `auth` module or back in the new `token-verifier` util module?

lahirumaramba replied:
Renamed back to `auth/token-verifier.ts`. I agree that it makes more sense now. Thanks!

hiranya911 commented on test/unit/auth/token-verifier.spec.ts:
```diff
@@ -45,6 +46,12 @@ const expect = chai.expect;
 const ONE_HOUR_IN_SECONDS = 60 * 60;
 const idTokenPublicCertPath = '/robot/v1/metadata/x509/securetoken@system.gserviceaccount.com';
 
+const AUTH_ERROR_CODE_CONFIG: ErrorCodeConfig = {
```
Why not reuse the const defined in the implementation?

lahirumaramba replied:
Updated to reuse the ones from the implementation. Thanks!

hiranya911 commented on test/unit/auth/token-verifier.spec.ts:
```diff
@@ -309,6 +326,34 @@ describe('FirebaseTokenVerifier', () => {
       }).to.throw('First argument to verifyIdToken() must be a Firebase ID token');
     });
 
+    const errorTypes = [
```
Why use an array with just one element?

lahirumaramba replied:
I was hoping that we will have other error types to include here in the future. Will remove the array for now and we can add it later when we have more if needed.

hiranya911 commented on test/unit/auth/token-verifier.spec.ts:
```diff
@@ -309,6 +326,34 @@ describe('FirebaseTokenVerifier', () => {
+      it('should throw with the correct error type and code set in token info', () => {
+        expect(() => {
+          (tokenVerifier as any).verifyJWT();
+        }).to.throw(errorType.type).with.property('code').match(
```
```equal(`auth/${errorType.config.invalidArg.code}`)```

lahirumaramba replied:
Updated!

hiranya911 reviewed (changes requested):
Looks pretty good. Just a few suggestions for cleaning up the PR.

lahirumaramba commented:
- Merged the expired error code to error code config
- Added validation for the new error code config and error type
- Added unit tests for new validations

hiranya911 commented on src/utils/token-verifier.ts:
```diff
@@ -213,9 +216,9 @@ export class FirebaseTokenVerifier {
             if (error.name === 'TokenExpiredError') {
               const errorMessage = `${this.tokenInfo.jwtName} has expired. Get a fresh ${this.tokenInfo.shortName}` +
-                ` from your client app and try again (auth/${this.tokenInfo.expiredErrorCode.code}).` +
+                ` from your client app and try again (auth/${this.tokenInfo.errorCodeConfig.expiredError.code}).` +
```
There's a reference to `auth` here. Should probably get rid of it.

lahirumaramba replied:
Oh good catch! I actually ended up finding more references to Auth. Ended up using generics in the verify token functions and moved the code to set the user id alias to one level up to `auth.ts`.
Let me know what you think. Thanks!

lahirumaramba replied:
Re: the error code prefix. I just removed the `auth` prefix, but the error code feels a bit redundant here as the error code will also be included in the public error type.

hiranya911 replied:
This is something that will get cleaned up once we fix error handling for this SDK. For now this solution is adequate.

hiranya911 commented on test/unit/auth/token-verifier.spec.ts:
```diff
@@ -30,10 +30,10 @@ import LegacyFirebaseTokenGenerator = require('firebase-token-generator');
 import { FirebaseTokenGenerator, ServiceAccountSigner } from '../../../src/auth/token-generator';
 import * as verifier from '../../../src/utils/token-verifier';
-import * as verifierUtil from '../../../src/auth/token-verifier-util';
+import * as verifierAuth from '../../../src/auth/token-verifier';
```
Easier and less ambiguous to just import named members here.

```
import { ID_TOKEN_INFO, SESSION_COOKIE_INFO } from '...';
```

lahirumaramba replied:
Updated!

hiranya911 commented on test/unit/auth/token-verifier.spec.ts:
```diff
@@ -301,13 +291,34 @@ describe('FirebaseTokenVerifier', () => {
-              expiredErrorCode: invalidExpiredErrorCode as any,
-              errorCodeConfig: AUTH_ERROR_CODE_CONFIG,
+              errorCodeConfig: verifierAuth.ID_TOKEN_ERROR_CODE_CONFIG,
+              errorType: invalidErrorCodeTypes as any,
```
Did you mean to pass `invalidErrorCodeType` here?

lahirumaramba replied:
Thanks! Fixed.

hiranya911 commented on test/unit/auth/token-verifier.spec.ts:
```diff
@@ -301,13 +291,34 @@ describe('FirebaseTokenVerifier', () => {
+      });
+    });
+
+    const invalidErrorCodeConfigs = [null, NaN, 0, 1, true, false, [], {}, { a: 1 }, _.noop, '', 'test'];
```
These tests now belong in utils.

lahirumaramba replied:
Moved the tests to `test/utils`

hiranya911 replied:
Let's not move the whole file. It still contains many auth-specific tests.

hiranya911 reviewed (changes requested):
Thanks for making the changes. Looks mostly good. Couple of points to revise:

1. There's one remaining reference to `auth` from the utils token verifier.
2. Unit tests could use a bit more attention.

lahirumaramba commented on src/auth/auth.ts:
```diff
@@ -520,8 +520,9 @@ export class BaseAuth<T extends AbstractAuthRequestHandler> implements BaseAuthI
-    return this.sessionCookieVerifier.verifyJWT(sessionCookie, isEmulator)
+    return this.sessionCookieVerifier.verifyJWT<DecodedIdToken>(sessionCookie, isEmulator)
       .then((decodedIdToken: DecodedIdToken) => {
+        decodedIdToken.uid = decodedIdToken.sub;
```
Moved the code to set the `uid` alias here. Another way is to add helper functions, `verifySessionCookie()` and `verifyIdToken()`, to `auth/token-verifier`. Let me know your thoughts.

lahirumaramba commented on src/utils/token-verifier.ts:
```diff
@@ -0,0 +1,290 @@
+   * @return {Promise<T>} A promise fulfilled with the decoded claims of the Firebase Auth ID
+   *                           token.
+   */
+  public verifyJWT<T extends object>(jwtToken: string, isEmulator = false): Promise<T> {
```
To tighten the type check we can introduce a base type for returned tokens...
```
interface FirebaseToken {
  // registered JWT claims
  iss: string
  aud: string
}

interface DecodedIdToken extends FirebaseToken {
...
}

public verifyJWT<T extends FirebaseToken>
```

Let me know your thoughts. Thanks!

hiranya911 replied:
Yeah, this looks good to me. But I don't think we need a generic at all. Just make auth extend the TokenVerifier class, and transform the FirebaseToken into DecodedIdToken there.

hiranya911 commented on test/unit/utils/token-verifier.spec.ts:
```diff
@@ -30,10 +30,12 @@ import LegacyFirebaseTokenGenerator = require('firebase-token-generator');
 import { FirebaseTokenGenerator, ServiceAccountSigner } from '../../../src/auth/token-generator';
 import * as verifier from '../../../src/utils/token-verifier';
-import * as verifierUtil from '../../../src/auth/token-verifier-util';
+import {
```
Sorry if I wasn't clear. But I didn't mean moving this entire file into utils. Some tests in this file are still very auth-specific, and therefore should be in auth. But the test cases that specifically test for the `utils.token-verifier`, should be moved into a new test module (like the `invalidErrorTypes` test cases).

I think a good rule of thumb is if a test requires an import from `auth`, then it probably belongs in `auth`.

hiranya911 reviewed (changes requested):
Lets find a more typesafe way to handle the returns. Plus separation of tests between auth and utils.

lahirumaramba commented:
Closing this as we decided to change the overall approach and design of the refactor. This work is replaced by #1204
